File: src/python/output_layer_2d.py
from output_neuron import OutputNeuron
from layer import Layer

class OutputLayer2D(Layer):
    """Shape-aware sink: output neurons expose a 2D frame snapshot per tick."""
    def __init__(self, height, width, smoothing):
        Layer.__init__(self, 0, 0, 0)
        self.height = int(height)
        self.width = int(width)
        self.frame = [[0.0 for _ in range(self.width)] for _ in range(self.height)]
        for row_idx in range(self.height):
            for col_idx in range(self.width):
                neuron = OutputNeuron(f"OUT[{row_idx},{col_idx}]", smoothing=smoothing)
                neuron.set_bus(self.get_bus())
                try:
                    neuron.owner = self
                except Exception:
                    pass
                self.get_neurons().append(neuron)

    def index(self, y, x):
        return int(y) * self.width + int(x)

    def propagate_from(self, source_index, value):
        if source_index < 0 or source_index >= len(self.get_neurons()):
            return
        neuron = self.get_neurons()[source_index]
        fired = neuron.on_input(value)
        if fired:
            neuron.on_output(value)

    def end_tick(self):
        # Update frame from each output neuron's last value, then decay
        for neuron_index, neuron in enumerate(self.get_neurons()):
            neuron.end_tick()
            row_idx = neuron_index // self.width
            col_idx = neuron_index % self.width
            self.frame[row_idx][col_idx] = neuron.get_output_value()
        self.get_bus().decay()

    def get_frame(self):
        return self.frame

File: src/java/ai/nektron/grownet/OutputLayer2D.java
package ai.nektron.grownet;

import java.util.List;

/**
 * Shape-aware output layer (e.g., image writer) using unified onInput/onOutput.
 */
public class OutputLayer2D extends Layer {

    private final int height;
    private final int width;
    private final double[][] frame;

    public OutputLayer2D(int height, int width, double smoothing) {
        // Base Layer constructs the lateral bus; no E/I/M neurons are created here.
        super(0, 0, 0);

        this.height = height;
        this.width  = width;
        this.frame  = new double[height][width];

        // Create one OutputNeuron per pixel.
        List<Neuron> neurons = getNeurons();
        for (int row = 0; row < height; ++row) {
            for (int col = 0; col < width; ++col) {

                neurons.add(new OutputNeuron("OUT[" + row + "," + col + "]",
                        getBus(),
                        SlotConfig.singleSlot()));

                // If your OutputNeuron ctor instead takes (id, bus, smoothing),
                // use this alternative:
                // neurons.add(new OutputNeuron("OUT[" + y + "," + x + "]", getBus(), smoothing));
            }
        }
    }

    /** Row-major index helper. */
    public int index(int row, int col) { return row * width + col; }

    /** Route a scalar into the output neuron with the given index. */

    public void propagateFrom(int sourceIndex, double value) {
        if (sourceIndex < 0 || sourceIndex >= getNeurons().size()) return;

        OutputNeuron outputNeuron = (OutputNeuron) getNeurons().get(sourceIndex);

        // Unified contract: bus factors are read inside the neuron.
        boolean fired = outputNeuron.onInput(value);
        if (fired) {
            outputNeuron.onOutput(value);
        }
    }

    /** End-of-tick housekeeping + snapshot the last output frame. */
    @Override
    public void endTick() {
        // Allow the layer bus to decay toward neutral this tick.
        super.endTick();

        for (int neuronIndex = 0; neuronIndex < getNeurons().size(); ++neuronIndex) {
            OutputNeuron outputNeuron = (OutputNeuron) getNeurons().get(neuronIndex);
            outputNeuron.endTick();

            int row = neuronIndex / width;
            int col = neuronIndex % width;
            frame[row][col] = outputNeuron.getOutputValue();
        }
    }

    /** Convenience for dashboards: last output image. */
    public double[][] getFrame() {
        return frame;
    }
}
